src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -176,7 +177,7 @@ public static JsonDocument Parse(Stream utf8Json, JsonDocumentOptions options =
             {
                 // Holds document content, clear it before returning it.
                 drained.AsSpan().Clear();
-                ArrayPool<byte>.Shared.Return(drained.Array);
+                ArrayPool<byte>.Shared.Return(drained.Array!);
```

Comment:
Instead of doing this, add `Debug.Assert(drained.Array != null);` right after the call to `ReadToEndAsync` on line 170 instead.